### DISCUSSION TITLE
Help with real world VRP - Soluntion seems random

### DISCUSSION THREAD

**Discussion by MrcLuna** [PyVRP/PyVRP#637] [Q&A] [answered]
Hi,

I'm trying to use pyVRP to solve a real-world VRP, but the solutions I am getting seem random. I'm primarily an R user and not very experienced with Python, so I am looking for some guidance on how to properly solve my problem.

For testing pyVRP, I have two GeoPackages containing the depot and clients I need to service. Since this is a real-world problem, I used OSMnx for my base graph, converting longitude and latitude to UTM to feed the pyVRP model. I also used OSMnx to get the real distance matrix, but the results were too chaotic. Therefore, I reverted to using Manhattan distances for testing.

The plotting of clients and depot seems correct:
![image](https://github.com/user-attachments/assets/d106d053-96e7-431f-ac4a-c32575ac0e39)

Clients and Depot visualized in GIS match what we see in the pyVRP model:
![image](https://github.com/user-attachments/assets/36e234d3-85cc-427b-96ef-8043dcad6695)

I'm trying to get up to 6 routes for 63 clients, using 6 vehicles with a capacity of 12 and assigning a pickup of 1 unit at each client.

Both Manhattan distances and distances calculated on the OSM graph result in 6 routes that zigzag all over the place, which is unexpected. I would expect some group formation. The result I got is shown below:

![image](https://github.com/user-attachments/assets/9ef2319d-4d98-47a9-b1a1-783a0d80ba8b)

When running on commercial routing software with similar parameters, the result is more as I expected:
![image](https://github.com/user-attachments/assets/394dbe2b-5010-4aab-a1e5-595f9389479d)

Therefore, I'm seeking guidance from the community on how to get a proper solution for my problem and what I'm doing wrong.  
The code I used is as follows:

```python
import osmnx as ox
import networkx as nx
import numpy as np
import geopandas as gpd
import gc
import matplotlib.pyplot as plt

gc.enable()

# Convert long, lat to UTM (meters)
import pyproj
def convert_coords(coords, src_crs='epsg:4326', dst_crs='epsg:31984'):
    transformer = pyproj.Transformer.from_crs(src_crs, dst_crs, always_xy=True)
    x_new, y_new = transformer.transform(coords[0], coords[1])
    return (int(x_new), int(y_new))


# Initialize model
from pyvrp import Model
m = Model()

# Vehicle type - 6 vehicles - 12 capacity
m.add_vehicle_type(6,12)

# Geopackages for depot and clients
clients_gdf = gpd.read_file("clientes.gpkg")
depot_gdf = gpd.read_file("ecoponto.gpkg")

# Open street map graph 
graph_file = "malha_Fortaleza.graphml"
graph = ox.load_graphml(graph_file)


# Function to get the nearest graph node from geopackage points
def find_nearest_node(graph, point):
    return ox.distance.nearest_nodes(graph, point[0], point[1])

# Get coords for the depot
depot_coords = (depot_gdf.geometry.x.iloc[0], depot_gdf.geometry.y.iloc[0])
depot_node = find_nearest_node(graph, depot_coords)
depot_coords = convert_coords(depot_coords)
depot = [{
    'id': -999,
    'coords': depot_coords,
    'node': depot_node,
    'demand': 0
}]


# Add depot to teh model
depot = m.add_depot(x=depot_coords[0], y=depot_coords[1])


# Get the clients coods and add to the model
client = [] 
clients = []

for idx, row in clients_gdf.iterrows():
    client_coords = (row.geometry.x, row.geometry.y)
    client_node = find_nearest_node(graph, client_coords)
    client = {
        'coords': convert_coords(client_coords),
        'node': client_node,
        'demand': 1 
    }
 
    clients = [
        m.add_client(x=client["coords"][0], y=client["coords"][1], pickup=client["demand"],
        delivery=0, service_duration=0)
    ]


# Verification plot
from pyvrp.plotting import plot_coordinates
_, ax = plt.subplots(figsize=(8, 8))
plot_coordinates(m.data(), ax=ax)

# Join depot and clients
locations = [depot] + clients

for frm in locations:
    for to in locations:
        distance = abs(frm.x - to.x) + abs(frm.y - to.y)  # Manhattan
        if frm.x == to.x and frm.y == to.y:
           distance = int(0)
        m.add_edge(frm, to, distance = distance)


# Solve VRP
from pyvrp.stop import MaxRuntime
res = m.solve(stop=MaxRuntime(10), display=True)  
print(res)

# Plot solution
from pyvrp.plotting import plot_solution
_, ax = plt.subplots(figsize=(8, 8))
plot_solution(res.best, m.data(), plot_clients=True, ax=ax)

```

Thanks. Any help is welcome.

**Comment by N-Wouda** [comment, marked as answer]
Hi @MrcLuna! 👋 

This
```python
    clients = [
        m.add_client(x=client["coords"][0], y=client["coords"][1], pickup=client["demand"],
        delivery=0, service_duration=0)
    ]
```
is suspicious: you're overwriting the `clients` list every time with the last added client. That suggests your `locations` list contains only the depot and the last client (can you check?). All other distances then default to a huge integer value, and your costs, and distance results in the output should be huge too (can you check?).

I suspect you meant to do
```python
    clients.append(
        m.add_client(x=client["coords"][0], y=client["coords"][1], pickup=client["demand"],
        delivery=0, service_duration=0)
    )
```
or simply use `m.locations`, rather than building a `locations` list yourself.

_Nb:_ it's _PyVRP_, not _pyVRP_.

**Reply by MrcLuna** [reply]
Thank you. That was essentially the problem, as I don't know Python. I'm trying to get a working example to explain to the IT team how to solve VRPTW and integrate it into our system. Thanks.

![image](https://github.com/user-attachments/assets/45457577-83d9-4671-a088-00ec62b0f2af)

**Reply by N-Wouda** [reply]
Great to hear!
